rust/kernel/net/mod.rs

```diff
@@ -0,0 +1,109 @@
+use core::mem;
+
+pub mod device;
+pub mod ethtool;
+pub mod netlink;
+pub mod rtnl;
+pub mod skbuff;
+
+#[doc(inline)]
+pub use macros::rtnl_link_ops;
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    // SAFETY: addr is a valid address pointer as above required, and has efficient unaligned access as KCONFIG stats
+    let a: u32 = unsafe { core::ptr::read_unaligned(addr as *const u32) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u32>() as u32) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u16 = unsafe { *(addr as *const u16) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u16>() as u16) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if give Ethernet address is all zeros.
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` must point to a valid ethernet address and is 16-bit aligned
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    // SAFETY: addr points to a valid ethernet address and is 16-bit aligned as required above.
+    unsafe { *(addr.cast::<u32>()) | *(addr.add(4).cast::<u16>()) as u32 == 0 }
+}
+
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    unsafe { *(addr as *const [u16; 3]) == [0; 3] }
+}
+
+///  is_valid_ether_addr - Determine if the given Ethernet address is valid.
+/// Check that the Ethernet address (MAC) is not 00:00:00:00:00:00, is not
+/// a multicast address, and is not FF:FF:FF:FF:FF:FF.
+///
+/// Return true if the address is valid.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
```

Comment:
Let's avoid "please note", since it is documentation anyway :-) 

Also, please use Markdown (everywhere, but specially here that it is rendered!) and leave an empty line in sections.